sdk/dynamicconfig/metric/controller/push/period_matcher.go

```diff
@@ -0,0 +1,130 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package push
+
+import (
+	"sync"
+	"time"
+
+	pb "github.com/open-telemetry/opentelemetry-proto/gen/go/experimental/metricconfigservice"
+	"go.opentelemetry.io/contrib/sdk/dynamicconfig/internal/metricpattern"
+	"go.opentelemetry.io/contrib/sdk/dynamicconfig/metric"
+)
+
+const tolerance float64 = 0.1
+
+type PeriodMatcher struct {
+	metrics   map[string]*CollectData
+	startTime time.Time
+
+	m     sync.Mutex
+	sched []*pb.MetricConfigResponse_Schedule
+}
+
+type CollectData struct {
+	lastCollected time.Time
+	period        time.Duration
+}
+
+func (matcher *PeriodMatcher) MarkStart(startTime time.Time) {
+	matcher.startTime = startTime
+}
+
+func (matcher *PeriodMatcher) ConsumeSchedules(sched []*pb.MetricConfigResponse_Schedule) {
+	matcher.m.Lock()
+	defer matcher.m.Unlock()
+
+	matcher.sched = sched
+	matcher.metrics = make(map[string]*CollectData)
+}
+
+// TODO: compute GCD for divisibility issues
+func (matcher *PeriodMatcher) GetMinPeriod() time.Duration {
```

Comment:
Also, is there simpler way to calculate this? We know that a response compliant to the protocol should have all periods divisible by the smallest period, then, how about just fetch the minimal period. Then, check all periods to ensure they are divisible by the smallest period, and return an error otherwise.

Also, I feel ConsumeSchedule and GetMinPeriod can be merged into a function called TryResetSchedules.

func TryResetSchedules(...) {
  min_period, error := getMinPeriod(schedule)
  if error != nil: return min_period, error
  matcher.m.Lock()
  deter matcher.m.Unlock()
  matcher.sched = sched
  ...
}

This way, the PeriodMatcher's API is simpler, and getMinPeriod does not need to acquire the lock.

Reply:
Interesting thought, the periods *should* be divisible by the smallest, but because there exists the option in the protocol to allow general periods, we've written in the specification that an SDK must nonetheless be able to support indivisible periods. It is a little extra complexity, though the overhead attaches to only the first time the SDK reads a schedule, and is likely not a bottleneck on the general export process.

I like the idea of combining `ConsumeSchedule` and `GetMinPeriod`. It shouldn't be hard to do, but my only hesitation is that would the API be less intuitive if we both set new schedules and find their minimum period in a single method? Is it natural to know that `TryResetSchedules` would return a min period?